compiler/packages/babel-plugin-react-compiler/src/Entrypoint/Gating.ts

```diff
@@ -21,13 +132,13 @@ export function insertGatedFunctionDeclaration(
   gating: NonNullable<PluginOptions['gating']>,
   referencedBeforeDeclaration: boolean,
 ): void {
-  if (referencedBeforeDeclaration) {
-    CompilerError.throwTodo({
-      reason: `Encountered a function used before its declaration, which breaks Forget's gating codegen due to hoisting`,
-      description: `Rewrite the reference to ${fnPath.node.id?.name ?? 'this function'} to not rely on hoisting to fix this issue`,
+  if (referencedBeforeDeclaration && fnPath.isFunctionDeclaration()) {
```

Comment:
Note that this changes us to only bail out for *function declarations` which are referenced before declaration.
See `invalid-fnexpr-reference` and `reassigned-fnexpr-variable` fixtures for examples why we don't need special handling for other function kinds.

This should be correct as other referenced-before-declaration identifiers are handled by directly replacing ast nodes. This may error at runtime, but still preserves source semantics (e.g. compiler-optimized code errors if and only if source code errors)